File: Atlassian/snake_python/snake_game.py
from board import Board
from coordinate import Coordinate
from snake import Snake

class SnakeGame:
    def __init__(self, m, n, snake_size):
        self.board = Board(m, n)
        self.snake = Snake(m // 2, n // 2, snake_size)
        self.moves_count = 0
        self.game_over = False

    def is_game_over(self):
        return self.game_over

    def move_snake(self, direction):
        if self.game_over:
            print("Game is already over!")
            return
        delta = self.new_move_delta(direction)
        self.moves_count += 1
        new_head = self.snake.get_head().add(delta)
        if not self.snake.is_valid_move(self.board.get_num_row(), self.board.get_num_col(), new_head):
            self.game_over = True
            print("Invalid Move, Game over")
            return
        snake_new_position = [new_head]
        current_position = self.snake.get_current_position()
        if self.moves_count % 5 == 0:
            snake_new_position.extend(current_position)
        else:
            snake_new_position.extend(current_position[:-1])
        self.snake.set_current_position(snake_new_position)
        self.snake.set_head(new_head)
        print(self.snake)

    def new_move_delta(self, direction):
        if direction == "left":
            return Coordinate(-1, 0)
        elif direction == "right":
            return Coordinate(1, 0)
        elif direction == "up":
            return Coordinate(0, -1)
        elif direction == "down":
            return Coordinate(0, 1)
        else:
            raise ValueError("Incorrect direction")

if __name__ == "__main__":
    # Create a 10x10 board with initial snake size of 3
    game = SnakeGame(10, 10, 3)
    # Test sequence of moves
    moves = ["right", "right", "down", "down", "left", "up"]
    for move in moves:
        print(f"\nMoving {move}")
        game.move_snake(move)
        if game.is_game_over():
            break 

File: Atlassian/AtlassianSnakeGame/src/main/java/org/codedesign/ops/SnakeGame.java
package org.codedesign.ops;

import org.codedesign.models.Board;
import org.codedesign.models.Coordinate;
import org.codedesign.models.Snake;

import javax.naming.directory.InvalidAttributesException;
import java.util.ArrayList;
import java.util.List;

public class SnakeGame {
    private Board board;
    private Snake snake;
    private Long movesCount;
    private boolean gameOver;

    public SnakeGame(int m, int n, int snakeSize) {
        board = new Board(m, n);
        snake = new Snake(m / 2, n / 2, snakeSize);
        movesCount = 0L;
        gameOver = false;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public void moveSnake(String direction) throws InvalidAttributesException {
        if (gameOver) {
            System.out.println("Game is already over!");
            return;
        }

        Coordinate delta = newMoveDelta(direction);
        movesCount = movesCount + 1;

        // Calculate new head position
        Coordinate newHead = snake.getHead().add(delta);

        // Check if move is valid
        if (!snake.isValidMove(board.getNumRow(), board.getNumCols(), newHead)) {
            gameOver = true;
            System.out.println("Invalid Move, Game over");
            return;
        }

        // Create new position list
        List<Coordinate> snakeNewPosition = new ArrayList<>();
        snakeNewPosition.add(newHead);

        // Add rest of body
        List<Coordinate> currentPosition = snake.getCurrentPosition();
        if (movesCount % 5 == 0) {
            // On growth turns, keep the tail
            snakeNewPosition.addAll(currentPosition);
        } else {
            // On normal turns, remove the tail
            for (int i = 0; i < currentPosition.size() - 1; i++) {
                snakeNewPosition.add(currentPosition.get(i));
            }
        }

        snake.setCurrentPosition(snakeNewPosition);
        snake.setHead(newHead);
        System.out.println(snake);
    }

    private Coordinate newMoveDelta(String direction) throws InvalidAttributesException {
        if (direction.equals("left")) {
            return new Coordinate(-1, 0);
        } else if (direction.equals("right")) {
            return new Coordinate(1, 0);
        } else if (direction.equals("up")) {
            return new Coordinate(0, -1);
        } else if (direction.equals("down")) {
            return new Coordinate(0, 1);
        } else {
            throw new InvalidAttributesException("Incorrect direction");
        }
    }

    public static void main(String[] args) {
        try {
            // Create a 10x10 board with initial snake size of 3
            SnakeGame game = new SnakeGame(10, 10, 3);

            // Test sequence of moves
            String[] moves = { "right", "right", "down", "down", "left", "up" };

            for (String move : moves) {
                System.out.println("\nMoving " + move);
                game.moveSnake(move);
                if (game.isGameOver()) {
                    break;
                }
            }

        } catch (InvalidAttributesException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}
